### PR TITLE
Pipelined Chain Sync Client

### PR DESCRIPTION
**Pull request by coot**
This PR depends on #960.

### CODE REVIEW

**Comment by coot** [comment]
@edsko I requested your review, since your team will use the `chain-sync-pipelined` client, so you might want to check out its API.

**Comment by avieth** [comment]
> This PR depends on #960.

But that PR says it was closed in favour of #950

File: ouroboros-consensus/test-consensus/Test/Consensus/ChainSyncClient.hs [hunk 1]
```diff
@@ -198,20 +202,22 @@ newtype ServerUpdates =
...
 -- syncing help generate different chains to start syncing from.
-runChainSync :: forall m.
+runChainSync :: forall m tip.
        ( MonadAsync m
        , MonadFork  m
        , MonadMask  m
        , MonadTimer m
        , MonadThrow (STM m)
+       , MonadSay   m
+       , tip ~ (Point (Header TestBlock), BlockNo)
```
Comment (avieth) [inline ouroboros-consensus/test-consensus/Test/Consensus/ChainSyncClient.hs]: Any downside to replacing `tip` with `(Point (Header TestBlock), BlockNo)` and removing the type parameter?

File: ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Test.hs [hunk 2]
```diff
@@ -294,6 +294,8 @@ prop_pipe_IO (TestChainAndPoints chain points) =
     ioProperty (prop_channel createPipeConnectedChannels chain points)
 
 
+-- TODO: issue #347: BlockFetch pipelined tests using channels & pipes
```
Comment (avieth) [inline ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Test.hs]: Are there no tests for this feature?

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+-- | Data received through pipelining: either roll forward or roll backward
+-- instcutrion.  If the server replied with 'MsgAwaitReply' the pipelined
+-- receiver will await for the next message which must come with an instruction
+-- how to update our chain.
+--
+-- Note: internal API, not exposed by this module.
+--
+data ChainSyncInstruction header tip
+    = RollForward  !header         !tip
+    | RollBackward !(Point header) !tip
```
Comment (avieth) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs]: Isn't the same thing defined in the non-pipelined peer? Possible to re-use?

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs]: no, there the client can use callbacks directly on received values; in the pipelined case we need to put them on a queue.

File: ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Test.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Test.hs]: That's right, as the comment suggests.

File: ouroboros-consensus/test-consensus/Test/Consensus/ChainSyncClient.hs (hunk 1 above)
Comment (coot) [reply ouroboros-consensus/test-consensus/Test/Consensus/ChainSyncClient.hs]: Downwards `tip` is used in two different independent places which could be substituted with different types and the code would compile, but if they don't agree the test would fail.
I think it's better to use `tip ~ ...` and when the `tip` will change modify it in one place without getting a false test failure.

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 4]
```diff
@@ -85,20 +86,20 @@ data ChainSyncClientException blk =
...
-    | InvalidRollBack (Point blk) (Point blk)
+    | InvalidRollBack (Point blk) tip
 
       -- | We send the upstream node a bunch of points from a chain fragment and
       -- the upstream node responded with an intersection point that is not on
       -- our chain fragment, and thus not among the points we sent.
       --
       -- We store the intersection point the upstream node sent us.
-    | InvalidIntersection (Point blk)
+    | InvalidIntersection (Point (Header blk))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: Why this change?
